posthog/models/team.py

```diff
@@ -139,6 +139,10 @@ def __str__(self):
 
 
 @receiver(models.signals.pre_delete, sender=Team)
-def team_deleted(sender, instance, **kwargs):
-    instance.event_set.all().delete()
-    instance.elementgroup_set.all().delete()
+def team_deleted(sender: Type[Team], instance: Team, **kwargs):
```

Comment:
There are other tables we have - e.g. session recordings, etc. What happens if user has rows in there, does the delete fail?

Reply:
So far we've only seen it get blocked on `posthog_person` a couple of times, so that's the main addition here, though I do see some potential for locking with session recordings and such too. I don't know what other than Postgres `ON DELETE CASCADE` could 100% eliminate this potential.

Reply:
I would assume team deletion fails if there's data in any of the tables that have team_id fk but aren't listed here.

Maybe the solution is just to do a CASCADE or...?

Reply:
Currently I have no better answer than recreating (`DROP` then `ADD`, `ALTER` doesn't work with `ON DELETE`) the problematic foreign constraints with `ON DELETE CASCADE`. Otherwise we have to trust that Django cleans up well by itself – which for the most part it does seem to do, only sometimes encountering what looks like a race condition.

Reply:
Okay. So say a team is deleted with rows in session_recording_events, would it after this change:

1. Blow up each time
2. Work but in a racy way due to django constraints?

Reply:
Tested this out by installing a custom plugin.

Project deletion failed for me:

```
Traceback (most recent call last):
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/db/models/fields/related_descriptors.py", line 173, in __get__
    rel_obj = self.field.get_cached_value(instance)
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/db/models/fields/mixins.py", line 15, in get_cached_value
    return instance._state.fields_cache[cache_name]
KeyError: 'person'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/rest_framework/viewsets.py", line 125, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/rest_framework/views.py", line 506, in dispatch
    response = handler(request, *args, **kwargs)
  File "/home/macobo/backups/2/projects/posthog/posthog/mixins.py", line 23, in destroy
    self.perform_destroy(instance)
  File "/home/macobo/backups/2/projects/posthog/posthog/mixins.py", line 13, in perform_destroy
    instance.delete()
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/db/models/base.py", line 947, in delete
    return collector.delete()
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/db/models/deletion.py", line 435, in delete
    signals.post_delete.send(
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/dispatch/dispatcher.py", line 177, in send
    return [
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/dispatch/dispatcher.py", line 178, in <listcomp>
    (receiver, receiver(signal=self, sender=sender, **named))
  File "/home/macobo/backups/2/projects/posthog/ee/clickhouse/models/person.py", line 48, in person_distinct_id_deleted
    delete_person_distinct_id(instance)
  File "/home/macobo/backups/2/projects/posthog/ee/clickhouse/models/person.py", line 124, in delete_person_distinct_id
    "person_id": person_distinct_id.person.uuid,
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/db/models/fields/related_descriptors.py", line 187, in __get__
    rel_obj = self.get_object(instance)
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/db/models/fields/related_descriptors.py", line 154, in get_object
    return qs.get(self.field.get_reverse_related_filter(instance))
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/db/models/query.py", line 429, in get
    raise self.model.DoesNotExist(
posthog.models.person.Person.DoesNotExist: Person matching query does not exist.
Internal Server Error: /api/projects/1/
Traceback (most recent call last):
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/db/models/fields/related_descriptors.py", line 173, in __get__
    rel_obj = self.field.get_cached_value(instance)
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/db/models/fields/mixins.py", line 15, in get_cached_value
    return instance._state.fields_cache[cache_name]
KeyError: 'person'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/rest_framework/viewsets.py", line 125, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/rest_framework/views.py", line 506, in dispatch
    response = handler(request, *args, **kwargs)
  File "/home/macobo/backups/2/projects/posthog/posthog/mixins.py", line 23, in destroy
    self.perform_destroy(instance)
  File "/home/macobo/backups/2/projects/posthog/posthog/mixins.py", line 13, in perform_destroy
    instance.delete()
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/db/models/base.py", line 947, in delete
    return collector.delete()
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/db/models/deletion.py", line 435, in delete
    signals.post_delete.send(
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/dispatch/dispatcher.py", line 177, in send
    return [
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/dispatch/dispatcher.py", line 178, in <listcomp>
    (receiver, receiver(signal=self, sender=sender, **named))
  File "/home/macobo/backups/2/projects/posthog/ee/clickhouse/models/person.py", line 48, in person_distinct_id_deleted
    delete_person_distinct_id(instance)
  File "/home/macobo/backups/2/projects/posthog/ee/clickhouse/models/person.py", line 124, in delete_person_distinct_id
    "person_id": person_distinct_id.person.uuid,
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/db/models/fields/related_descriptors.py", line 187, in __get__
    rel_obj = self.get_object(instance)
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/db/models/fields/related_descriptors.py", line 154, in get_object
    return qs.get(self.field.get_reverse_related_filter(instance))
  File "/home/macobo/.pyenv/versions/posthog-core/lib/python3.8/site-packages/django/db/models/query.py", line 429, in get
    raise self.model.DoesNotExist(
posthog.models.person.Person.DoesNotExist: Person matching query does not exist.
```

Reply:
What was the plugin?

Reply:
Don't remember, random one from the library.

Reply:
As for the error trace you posted, it seems that using raw `DELETE` queries in `pre_delete` messes with Django ORM's caching (i.e. the actual delete stage tries to delete rows that already were deleted in pre-delete), so I reverted back to QuerySet deletion.
I also made it so that this test uses `create_demo_team`, which does create a bunch of varied demo data (incl. session recordings).
However that does not reproduce the actual problem: the race condition. I'm not sure how to proceed with that.
Maybe we should go for the "atomic" option: making sure problematic foreign keys actually have `ON DELETE CASCADE` in the DB, instead of having Django try to replicate the behavior imperfectly?

Reply:
I think `ON DELETE CASCADE` is reasonable as long as we make sure nothing is unintentionally deleted when that is executed.